### PR TITLE
Split incoming/outgoing packet registry, transition protocol states correctly

### PR DESCRIPTION
**Pull request by AlexProgrammerDE**
These changes are necessary to avoid race-conditions in the code and to prepare the codebase for ViaVersion support.

### CODE REVIEW

**Comment by AlexProgrammerDE** [comment]
PR is ready for review

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: Let's add nonnull/nullable annotations here

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: ?
seems incomplete

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/packet/Packet.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/packet/Packet.java]:
```suggestion
     * Returns whether the packet is terminal. If true, this should be the last packet sent inside a protocol state.
     * Subsequently, {@link Session#setAutoRead(boolean)} should be disabled when a terminal packet is received, until the session has switched into a new state and is ready to receive more packets.
     *
     * @return Whether the packet is terminal.
```
probably messed up the link formatting, and requires import

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: Nullable annotation on `onSent`

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: last sentence is cutoff

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: javadoc, and i'd suggest `switchInboundState` because thats the format the getters/setters use

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpFlowControlHandler.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpFlowControlHandler.java:18]: comment about why this is necessary

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ClientListener.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ClientListener.java]: looks like the inboundState switch and intention send can be pulled out of both cases, out of the switch

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: can we not check for `ClientboundStartConfigurationPacket` in `packetSent`, and setOutboundState to CONFIGURATION there? `packetSent` should be invoked more or less right after the `onSent` callback would be?

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: are we gonna end up with an amount of threads equal to the how many configuration phases we've gone through?

**Review by Konicai — DISMISSED** [review]
nice, especially like the cleanup in the listeners

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: The "proper" way to handle keep-alive is to reset the keepalive listener of the session during protocol switching.

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: I'd rather have the developers explicitly switch them like in vanilla. The entire onSent thing is made because of that reason. Stuff in packetSent may be not be handled immediately because packets sometimes get handled on the event loop. However onSent is guaranteed to be handled inside the netty pipeline.

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: Maybe I should rework this a bit

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: ClientboundStartConfigurationPacket is not isPriority
therefore it gets handled on the event loop 
so if you 
send(ClientboundStartConfigurationPacket)
and
send(SomeConfigurationPacket)
as a developer right after each other, it'll error because the event loop is having a race condition with the netty pipeline 
One other fix would be setting ClientboundStartConfigurationPacket as isPriority, but I don't wanna touch that with this PR
onSent allows the dev to change the state the next packets gets decoded with immediately without the chance of race conditions because of off-pipeline processing

Comment (Konicai) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: I'm sorry if I don't understand.

> Stuff in packetSent may be not be handled immediately because packets sometimes get handled on the event loop. However onSent is guaranteed to be handled inside the netty pipeline.

I read them as being called right after eachother? https://github.com/AlexProgrammerDE/MCProtocolLib/blob/2b310d2618f23127640a85194623126b1eef594b/protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpSession.java#L271-L275

Also, isPriority doesn't seem to apply to outbound packets?

Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: Ah yeah you're right. I was thinking too much from the clients perspective that all packets are in the event loop. Would still be weird though to add back packetSent... 
It makes it harder to compare logic with vanilla.

Comment (Konicai) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: It does have the distinction now that any packets handled by ServerListener in the packetSent event are packets that have been sent by the library user, not internal MCPL. I think its worth it to not introduce state switching to the library user for a single case (if i understand correctly)

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]:
```suggestion
     * @param autoRead Whether to enable auto read.
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: I wonder if we should just debug print a packet once, otherwise there's 2 debug lines for every packet.. might be a bit overkill 

If the decoding fails, it would print an error with the packet id anyways

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java]:
```suggestion
        log.debug("Setting inbound protocol state to: {}", state);
```
Might be just a bit clearer

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java]:
```suggestion
        log.debug("Setting outbound protocol state to: {}", state);
```
Same as above

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:291]: Is this a TODO for this PR?

**Review by onebeastchris — APPROVED** [review]

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:291]: No, not part of this PR.

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: Was fixed 3 weeks ago.

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: Was fixed 3 weeks ago.

Comment (onebeastchris) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: Packet is still not marked, which I'd make NonNull

Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: done

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: This is now not even possible anymore since explicit flushes after the writeAndFlush are needed due to race conditions. The code doesn't use onSent anymore for state switching. Developers need to manually switch state after sending the packet.

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: True, but imo it should be kept like this in case of for example deadlocks. If the deserialization never returns a result or an error and instead infitely blocks the thread we'd have no idea what is going on.

Comment (onebeastchris) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: If it returns an error, that would be logged already, no? 

My concern would be flooding debug logs with double the lines for just a little more information that is, more or less, redundant

Comment (Camotoy) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: If we really want this, can we set one to `fine`/`trace` logging level?

Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: Yeah ig trace works

**Review by Redned235 — APPROVED** [review]

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: Done

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java]:
```suggestion
                    pipeline.addLast("flow-control", new TcpFlowControlHandler());
                    pipeline.addLast("codec", new TcpPacketCodec(TcpClientSession.this, true));
                    pipeline.addLast("manager", TcpClientSession.this);
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpSession.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpSession.java:228]: comment please about this being java client behavior

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ClientListener.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ClientListener.java:146]: nice fix, should resolve issues with polymer

**Review by onebeastchris — APPROVED** [review]
lgtm

**Comment by AlexProgrammerDE** [comment]
Race conditions and weird disconnects are fixed. This PR can be merged.

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java]: ?

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpServer.java]:
```suggestion
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java]:
```suggestion
```

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: do we need the trace enabled check? doesn't log.trace() perform it too before formatting the string?

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/DropPacketFilter.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/DropPacketFilter.java]: I understand why it's named filter, but later in the pipeline it's referred to as "packet_filter" which imo doesn't quite convey the functionality of the pipeline "flushing" (why'd you rename it to dropping anyways? Just curious)

Maybe `FlushHandler` or something?

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpPacketCodec.java]: This is to avoid array allocations and expensive method calls, packet.getClass().getSimpleName() will lag a tiny bit and the trace enabled check prevents it being called.

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/DropPacketFilter.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/DropPacketFilter.java]: Yeah idk about the inconsistency, I have renamed and replaced this thing like 5 times until I got this working impl. I'll rename it to FlushHandler.

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (Konicai) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: > Would still be weird though to add back packetSent...
It makes it harder to compare logic with vanilla.

Minecraft is also the maintainer AND user of its network code—I think it is worth it here for the sake of ease of use.

Minecraft has a method for sending ClientboundStartConfigurationPacket and switching outbound protocol, but I'm not sure this project's structure would facilitate something like that?

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: What specifically is this being cached to a variable for? The field value seems to be mutated when updating the keep alive, but replaced on login and config start/end.

If we are trying to prevent races with the keep alive thread, have we considered making `KeepAliveState` immutable?

Also, could `keepAlivePending` be replaced with the field being null when not pending, and have the field be volatile?

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: This code is fairly long and shared between the two states, would be good to move to a method

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: This is coded closer to mojangs impl.
The pending field and other fields are part of the commonpacketlistener. And each protocol state keeps track of separate keepalives.

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: I'd target a rewrite that more closely resembles mojangs impl where each protocol state listener extends a common listener with common logic. I'd not worry too much about duplicated code rn.

Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: done

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java
Comment (Camotoy) [inline protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java]: Seems like we could just have a setState/setStates method just for convenience.

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java
Comment (onebeastchris) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/tcp/TcpClientSession.java]: Please use the NonNull annotation from the checkerframework lib

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java]: Is this really necessary? The only real reason for the protocol to be set for both at once is for handshake. There's currently no other reason for setting both inbound and outbound at once. There are also separate operations on the Session used for switching protocols.

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java]: Changing it now would make diffing it with vanilla harder. Also, if I restructure it, there's the possibility something I code doesn't properly match up with vanilla.

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/ServerListener.java:183]: Could you elaborate? 
What I'm saying is that the calling thread for the switchOutbound matters. Always forcing it to be inside netty threads could lead to deadlocks if the event loop is used for the inbound processing.

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java
Comment (Camotoy) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java]: It's done three times here. Might as well have a private method for it?

Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java]: We could have a private method to reset current state to handshake. Will add that tomorrow.

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java:287]: this doesn't need to be exposed if `switchOutboundState` is implemented in TcpSession

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (Konicai) [inline protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java]: grammar

File: protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java:287]: Yes, this is new API. It's already public for TcpSession. The reason why I'm exposing it is so it's easier to make ViaVersion integrations in the future as it needs direct channel access.

Comment (Konicai) [reply protocol/src/main/java/org/geysermc/mcprotocollib/network/Session.java:287]: sounds good 👍

File: protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java
Comment (AlexProgrammerDE) [reply protocol/src/main/java/org/geysermc/mcprotocollib/protocol/MinecraftProtocol.java]: done

**Comment by onebeastchris** [comment]
As discussed on discord, being able to disable/suppress the packet en/decoding logging while in debug mode using e.g. markers as flags would be ideal.
Alternatively, the logging addition could get implemented in a separate PR to not block this PR further - IMO it's the only thing remaining

**Comment by AlexProgrammerDE** [comment]
Should be able to be reviewed now. Added the markers.

**Pull request merged**
